File: C/4/13.c
#include <stdio.h>
#include <string.h>

/**
 * @brief Reverses a string s in place.
 *
 * @param s The string to be reversed.
 */
void reverse (char s[])
{
    static int begin = 0; /* Index of start of the string s. */
    static int end = 1;   /* Index of end of the string s. */

    /* If we are in the first run of this function */
    if (begin == 0) {
        end = strlen(s) - 1; /* Compute string size, excluding null character */
    }

    /* If begin overlap end, we reached base case. */
    if (begin >= end) {
        /* Resetting static variables */
        begin = 0;
        end = 1;
    } else {
        int _c = 0; /* Temporary variable to perform the swap. */

        /* Swap values */
        _c = s[begin];
        s[begin++] = s[end];
        s[end--] = _c;

        /* Call reverse recursively */
        reverse(s);
    }
}

main()
{
    char first[] = "Socorram-me, subi no onibus em Marrocos!";
    char second[] = "ovo";
    char third[] = "Lorem ipsum";

    printf("[*]\"%s\" =>\n", first);
    reverse(first);
    printf("\t\t\"%s\"\n\n", first);

    printf("[*]\"%s\" =>\n", second);
    reverse(second);
    printf("\t\t\"%s\"\n\n", second);

    printf("[*]\"%s\" =>\n", third);
    reverse(third);
    printf("\t\t\"%s\"\n\n", third);
}

File: rustlang/4/13.rs
/// @brief Reverses a string s in place.
///
/// @param s The string to be reversed.
unsafe fn reverse (s: &mut String)
{
    static mut BEGIN: usize = 0; // Index of start of the string s.
    static mut AUX: usize = 0;   // Used to compute the end of the string s.

    let end = s.len() - AUX; // Compute current end position

    if BEGIN < end-1 {

        // Remove character from current end
        let mut _b = if end < s.len() { s.remove(end-1) } else { s.pop().unwrap_or('#') };

        // Insert at current beginning
        s.insert(BEGIN, _b);

        // Remove character from current begin
        let mut _e = s.remove(BEGIN+1);

        // Insert at current end
        if end <= s.len() { s.insert(end-1, _e); } else { s.push(_e); }

        // Update begin and offset
        BEGIN += 1;
        AUX += 1;

        // Call reverse recursively
        reverse(s);
    } else {
        // Resetting static variables
        BEGIN = 0;
        AUX = 0;
    }
}

fn main()
{
    let mut first = String::from("Socorram-me, subi no onibus em Marrocos!");
    let mut second = String::from("ovo");
    let mut third = String::from("Lorem ipsum");

    unsafe {
        print!("[*]\"{}\" =>\n", first);
        reverse(&mut first);
        print!("\t\t\"{}\"\n\n", first);

        print!("[*]\"{}\" =>\n", second);
        reverse(&mut second);
        print!("\t\t\"{}\"\n\n", second);

        print!("[*]\"{}\" =>\n", third);
        reverse(&mut third);
        print!("\t\t\"{}\"\n\n", third);
    }
}
